File: apple/InlineKit/Sources/RealtimeV2/Sync/SyncStorage.swift
import Foundation
import InlineProtocol

public struct SyncState: Sendable {
  public let lastSyncDate: Int64

  public init(lastSyncDate: Int64) {
    self.lastSyncDate = lastSyncDate
  }
}

public struct BucketState: Sendable {
  public let date: Int64
  public let seq: Int64

  public init(date: Int64, seq: Int64) {
    self.date = date
    self.seq = seq
  }
}

public enum BucketKey: Sendable, Hashable {
  case space(id: Int64)
  case chat(peer: InlineProtocol.Peer)
  case user

  public func getEntityId() -> Int64 {
    switch self {
      case let .space(id):
        id

      case let .chat(peer):
        switch peer.type {
          case let .chat(value):
            -1 * value.chatID
          case let .user(value):
            value.userID
          default:
            fatalError("Invalid peer type")
        }

      // there is only one user and that's ours
      case .user:
        0
    }
  }

  /// In sync with backend
  public func getBucket() -> Int {
    switch self {
      case .chat:
        1
      case .user:
        2
      case .space:
        3
    }
  }

  public func toProtocolBucket() -> InlineProtocol.UpdateBucket {
    switch self {
      case let .chat(peer):
        InlineProtocol.UpdateBucket.with { $0.chat = .with { $0.peerID = peer.toInputPeer() } }
      case let .space(id):
        InlineProtocol.UpdateBucket.with { $0.space = .with { $0.spaceID = id } }
      case .user:
        InlineProtocol.UpdateBucket.with { $0.user = .init() }
    }
  }
}

public protocol SyncStorage: Sendable {
  func getState() async -> SyncState
  func setState(_ state: SyncState) async

  func getBucketState(for key: BucketKey) async -> BucketState
  func setBucketState(for key: BucketKey, state: BucketState) async

  /// Uses a single transaction
  func setBucketStates(states: [BucketKey: BucketState]) async

  /// Clears global sync state and all bucket states.
  func clearSyncState() async
}

File: web/packages/client/src/realtime/sync/sync-storage.ts
import type { SyncBucketCursor, SyncBucketKey, SyncState } from "./sync-types"
import { syncBucketId } from "./sync-types"

export interface SyncStorage {
  initialize(): Promise<void>
  getState(): Promise<SyncState>
  setState(state: SyncState): Promise<boolean>
  getBucketState(key: SyncBucketKey): Promise<SyncBucketCursor>
  setBucketState(key: SyncBucketKey, state: SyncBucketCursor): Promise<boolean>
  commitBucketState?(
    key: SyncBucketKey,
    state: SyncBucketCursor,
    apply: () => void,
  ): Promise<boolean>
  removeBucketState(key: SyncBucketKey): Promise<boolean>
  clear(): Promise<boolean>
}

export class MemorySyncStorage implements SyncStorage {
  private state: SyncState = { lastSyncDate: 0 }
  private readonly buckets = new Map<string, SyncBucketCursor>()

  async initialize() {}

  async getState() {
    return this.state
  }

  async setState(state: SyncState) {
    this.state = state
    return true
  }

  async getBucketState(key: SyncBucketKey) {
    return this.buckets.get(syncBucketId(key)) ?? { date: 0, seq: 0 }
  }

  async setBucketState(key: SyncBucketKey, state: SyncBucketCursor) {
    this.buckets.set(syncBucketId(key), state)
    return true
  }

  async removeBucketState(key: SyncBucketKey) {
    this.buckets.delete(syncBucketId(key))
    return true
  }

  async clear() {
    this.state = { lastSyncDate: 0 }
    this.buckets.clear()
    return true
  }
}
